src/graph/LdapAuthenticator.cpp

```diff
@@ -0,0 +1,362 @@
+/* Copyright (c) 2020 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+#include "graph/LdapAuthenticator.h"
+#include "graph/GraphFlags.h"
+#define LDAP_DEPRECATED 1
+
+namespace nebula {
+namespace graph {
+
+Status LdapAuthenticator::prepare() {
+    if (FLAGS_ldap_server.empty()) {
+        return Status::Error("LDAP authentication \"ldap_server\" parameter is not set.");
+    }
+
+    if (FLAGS_ldap_scheme.compare("ldap")) {
+        return Status::Error("LDAP authentication ldap_scheme only support \"ldap\".");
+    }
+
+    if (FLAGS_ldap_port == 0) {
+        FLAGS_ldap_port = ldapPort_;
+    }
+
+    // Simple bind mode
+    if (!FLAGS_ldap_prefix.empty() || !FLAGS_ldap_suffix.empty()) {
+        if (!FLAGS_ldap_basedn.empty() ||
+            !FLAGS_ldap_binddn.empty() ||
+            !FLAGS_ldap_bindpasswd.empty() ||
+            !FLAGS_ldap_searchattribute.empty() ||
+            !FLAGS_ldap_searchfilter.empty()) {
+            return Status::Error("LDAP authentication direct bind mode and search bind mode"
+                                 " parameters cannot be mixed.");
+        }
+    // Here search bind mode must be set FLAGS_ldap_basedn
+    } else if (FLAGS_ldap_basedn.empty()) {
+        return Status::Error("LDAP authentication requires argument \"ldap_prefix\", "
+                             "\"ldap_suffix\", or \"ldap_basedn\" to be set.");
+    }
+
+    /**
+     * Search bind mode can either use FLAGS_ldap_searchattribute or
+     * FLAGS_ldap_searchfilter, but can't use both. FLAGS_Ldap_searchattribute
+     * default is "uid". FLAGS_Ldap_searchfilter is more flexible search filtes
+     * than FLAGS_Ldap_searchattribute.
+     */
+    if (!FLAGS_ldap_searchattribute.empty() && !FLAGS_ldap_searchfilter.empty()) {
+        return Status::Error("LDAP authentication cannot use ldap_searchattribute "
+                             "together with ldap_searchfilter.");
+    }
+
+    return Status::OK();
+}
+
+
+Status LdapAuthenticator::initLDAPConnection() {
+    int ldap_version = LDAP_VERSION3;
+
+    // Ldap_create interface is only available in versions higher than 2.4.49
+    /*
+    int ret  = ldap_create(&ldap_);
+    if (ret != LDAP_SUCCESS) {
+        return Status::Error("Init LDAP failed.");
+    }
+    ldap_->ld_options.ldo_defport = FLAGS_ldap_port;
+
+    ret = ldap_set_option(ldap_, LDAP_OPT_HOST_NAME, FLAGS_ldap_server);
+    if (ret != LDAP_SUCCESS) {
+        ldap_ld_free(ldsp_, 1, NULL, NULL);
+        return Status::Error("Init LDAP failed.");
+    }
+    */
+    // ldap_init is deprecated, so use ldap_initialize
```

Comment:
not deleted comments ?

Reply:
I actually prefer the ldap_create method here.
`E.g. ldap_create (& ldap_);`
But ldap_create requires version 2.4.49 and above.

  `ldap_init (FLAGS_ldap_server.c_str (), FLAGS_ldap_port);`
ldap_init is deprecated. If you use it, you need to add` -DLDAP_DEPRECATED` in  CMakeLists.txt , so it is not recommended to use ldap_init.

I am also hesitant to remove the comments. I put this note here to remind us that the above two methods cannot be used now.